rectools/models/bert4rec.py

```diff
@@ -0,0 +1,197 @@
+    PointWiseFeedForward,
+    PositionalEncodingBase,
+    SessionEncoderDataPreparatorBase,
+    SessionEncoderLightningModule,
+    SessionEncoderLightningModuleBase,
+    TransformerLayersBase,
+    TransformerModelBase,
+)
+
+PADDING_VALUE = "PAD"
+MASKING_VALUE = "MASK"
+
+
+class BERT4RecDataPreparator(SessionEncoderDataPreparatorBase):
+    """TODO"""
+
+    def __init__(
+        self,
+        session_max_len: int,
+        batch_size: int,
+        dataloader_num_workers: int,
+        train_min_user_interactions: int,
+        mask_prob: float,
+        item_extra_tokens: tp.Sequence[tp.Hashable],
+        shuffle_train: bool = True,
+    ) -> None:
+        super().__init__(
+            session_max_len=session_max_len,
+            batch_size=batch_size,
+            dataloader_num_workers=dataloader_num_workers,
+            train_min_user_interactions=train_min_user_interactions,
+            item_extra_tokens=item_extra_tokens,
+            shuffle_train=shuffle_train,
+        )
+        self.mask_prob = mask_prob
+
+    def _mask_session(self, ses: List[int]) -> Tuple[List[int], List[int]]:
+        masked_session = ses.copy()
+        target = ses.copy()
+        random_probs = np.random.rand(len(ses))
+        for j in range(len(ses)):
+            if random_probs[j] < self.mask_prob:
+                random_probs[j] /= self.mask_prob
+                if random_probs[j] < 0.8:
+                    masked_session[j] = 1
+                elif random_probs[j] < 0.9:
+                    masked_session[j] = np.random.randint(low=2, high=self.item_id_map.size, size=1)[0]
+            else:
+                target[j] = 0
+        return masked_session, target
+
+    def _collate_fn_train(
+        self,
+        batch: List[Tuple[List[int], List[float]]],
+    ) -> Tuple[torch.LongTensor, torch.LongTensor, torch.FloatTensor]:
+        """TODO"""
+        batch_size = len(batch)
+        x = np.zeros((batch_size, self.session_max_len))
+        y = np.zeros((batch_size, self.session_max_len))
+        yw = np.zeros((batch_size, self.session_max_len))
+        for i, (ses, ses_weights) in enumerate(batch):
+            masked_session, target = self._mask_session(ses)
+            x[i, -len(ses) + 1 :] = masked_session[:-1]  # ses: [session_len] -> x[i]: [session_max_len]
+            y[i, -len(ses) + 1 :] = target[:-1]  # ses: [session_len] -> y[i]: [session_max_len]
+            yw[i, -len(ses) + 1 :] = ses_weights[:-1]  # ses_weights: [session_len] -> yw[i]: [session_max_len]
+
+        return torch.LongTensor(x), torch.LongTensor(y), torch.FloatTensor(yw)
+
+    def _collate_fn_recommend(self, batch: List[Tuple[List[int], List[float]]]) -> torch.LongTensor:
+        """Right truncation, left padding to session_max_len"""
+        x = np.zeros((len(batch), self.session_max_len))
+        for i, (ses, _) in enumerate(batch):
+            session = ses.copy()
+            session = session[1:] + [1]
```

Comment:
What is sessions had only two items? Why are we dropping the first one on inference?

Reply:
fixed